### PR TITLE
Added way to remove filters from channel stack builder

### PR DESCRIPTION
**Pull request by dgquintas**
This would allow different implementations of the same plugin filter (as indicated by using the same filter name) to be registered. Plugin init function are able to remove pre-existing versions of the filter (as determined by having the same filter name) as part of their registration process.

### CODE REVIEW

File: src/core/lib/channel/channel_stack.h [hunk 1]
```diff
@@ -159,6 +159,11 @@ typedef struct {
 
   /* The name of this filter */
   const char *name;
+
+  /* For filter with the same \a name, determines which one is kept in the
+   * channel stack. Stringly lesser values take precedence (UNIX priority
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack.h]: What does "Stringly" mean in this context?  Does it mean that we're printing the values as strings and then doing string comparisons?  If so, why?  Isn't it both easier and more efficient to just do a numeric comparison?

File: src/core/lib/channel/channel_stack_builder.h [hunk 2]
```diff
@@ -116,15 +120,29 @@ bool grpc_channel_stack_builder_add_filter_before(
...
     grpc_post_filter_create_init_func post_init_func,
     void *user_data) GRPC_MUST_USE_RESULT;
 
-/// Add \a filter to the end of the filter list.
+/// Add \a filter to the end of the filter list, unless a filter by the
+/// same name already exists. In that case, the one with a higher numerical
+/// priority value will be replaced in its current place.
+/// Call \a post_init_func(..., \a user_data) once the channel stack is
+/// created.
+bool grpc_channel_stack_builder_append_filter(
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.h]: Looks like there are now two identical declarations of the same function here.

File: src/core/lib/channel/channel_stack_builder.h [hunk 3]
```diff
@@ -95,6 +95,10 @@ bool grpc_channel_stack_builder_move_next(
 bool grpc_channel_stack_builder_move_prev(
     grpc_channel_stack_builder_iterator *iterator);
 
+/// Return an iterator at \a filter_name, or NULL if not found.
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.h]: To make this interface work more like the C++ STL iterator interface, would it make sense to have this return an iterator pointing to the end instead of NULL if the name is not found?

File: src/core/lib/channel/channel_stack_builder.c [hunk 4]
```diff
@@ -158,9 +176,43 @@ const grpc_channel_args *grpc_channel_stack_builder_get_channel_arguments(
...
+static bool search_and_maybe_replace(
+    grpc_channel_stack_builder *builder, const grpc_channel_filter *filter,
+    grpc_post_filter_create_init_func post_init_func, void *user_data) {
+  grpc_channel_stack_builder_iterator *it =
+      grpc_channel_stack_builder_iterator_find(builder, filter->name);
+  if (it == NULL) {
+    return false;
+  }
+  if (filter->priority < it->node->filter->priority) {
+    // If the new filter has a lesser priority value (ie, it's more prioritary)
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.c]: I don't think "prioritary" is actually an English word. :)

File: src/core/lib/channel/channel_stack_builder.c [hunk 5]
```diff
@@ -158,9 +176,43 @@ const grpc_channel_args *grpc_channel_stack_builder_get_channel_arguments(
...
+    // than the present one, replace.
+    filter_node *new_node = gpr_malloc(sizeof(*new_node));
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.c]: Couldn't this entire block just be replaced with the following?

```
it->node->filter = filter;
it->node->init = post_init_func;
it->node->init_arg = user_data;
```

File: src/cpp/common/channel_filter.h [hunk 6]
```diff
@@ -386,7 +388,7 @@ void RegisterChannelFilter(
        FilterType::SetPollsetOrPollsetSet, FilterType::DestroyCallElement,
        FilterType::channel_data_size, FilterType::InitChannelElement,
        FilterType::DestroyChannelElement, FilterType::GetPeer,
-       FilterType::GetChannelInfo, name}};
+       FilterType::GetChannelInfo, name, INT_MAX}};
```
Comment (markdroth) [inline src/cpp/common/channel_filter.h]: Instead of hard-coding this, let's add a new parameter to `RegisterChannelFilter()`.  Otherwise, there's no way for someone internally to define a C++ filter that overrides one of the OSS filters.

File: src/core/lib/channel/channel_stack.h (hunk 1 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack.h]: I have no idea where that word came from. It's the result of a race condition in my brain. It should be **Strictly**, but I think my muscle memory is more used to writing "string" when a word begins with "stri".

File: src/core/lib/channel/channel_stack_builder.h (hunk 3 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack_builder.h]: Done.

File: src/core/lib/channel/channel_stack_builder.h (hunk 2 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack_builder.h]: Not applicable anymore.

File: src/core/lib/channel/channel_stack_builder.c (hunk 4 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack_builder.c]: Not applicable anymore, but point taken :)

File: src/core/lib/channel/channel_stack_builder.c (hunk 5 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack_builder.c]: Not applicable anymore.

File: src/cpp/common/channel_filter.h (hunk 6 above)
Comment (dgquintas) [inline src/cpp/common/channel_filter.h]: Not applicable anymore.

**Comment by dgquintas** [comment]
@markdroth I've implemented one of your suggestions. PTAL.

File: src/core/lib/channel/channel_stack_builder.h [hunk 7]
```diff
@@ -95,6 +95,11 @@ bool grpc_channel_stack_builder_move_next(
 bool grpc_channel_stack_builder_move_prev(
     grpc_channel_stack_builder_iterator *iterator);
 
+/// Return an iterator at \a filter_name, or at the end of the iterator if not
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.h]: s/end of the iterator/end of the list/

File: src/core/lib/channel/channel_stack_builder.c [hunk 8]
```diff
@@ -124,6 +124,24 @@ bool grpc_channel_stack_builder_move_prev(
...
+grpc_channel_stack_builder_iterator *grpc_channel_stack_builder_iterator_find(
+    grpc_channel_stack_builder *builder, const char *filter_name) {
+  GPR_ASSERT(filter_name != NULL);
+  grpc_channel_stack_builder_iterator *it =
+      grpc_channel_stack_builder_create_iterator_at_first(builder);
+  if (grpc_channel_stack_builder_iterator_is_end(it)) return it;
+  do {
+    const char *filter_name_at_it =
+        grpc_channel_stack_builder_iterator_filter_name(it);
+    if (filter_name_at_it != NULL &&
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.c]: Is it actually possible for the filter name to be NULL?

File: src/core/lib/channel/channel_stack_builder.c [hunk 9]
```diff
@@ -124,6 +124,24 @@ bool grpc_channel_stack_builder_move_prev(
   return true;
 }
 
+grpc_channel_stack_builder_iterator *grpc_channel_stack_builder_iterator_find(
+    grpc_channel_stack_builder *builder, const char *filter_name) {
+  GPR_ASSERT(filter_name != NULL);
+  grpc_channel_stack_builder_iterator *it =
+      grpc_channel_stack_builder_create_iterator_at_first(builder);
+  if (grpc_channel_stack_builder_iterator_is_end(it)) return it;
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.c]: I think the logic here would be a bit clearer if written like this:

```
for (; !grpc_channel_stack_builder_iterator_is_end(it);
     grpc_channel_stack_builder_move_next(it)) {
  const char *filter_name_at_it =
      grpc_channel_stack_builder_iterator_filter_name(it);
  if (filter_name_at_it != NULL &&
      strcmp(filter_name, filter_name_at_it) == 0) {
    break;
  }
}
return it;
```

File: src/core/lib/channel/channel_stack_builder.h [hunk 10]
```diff
@@ -132,6 +137,11 @@ bool grpc_channel_stack_builder_append_filter(
     grpc_post_filter_create_init_func post_init_func,
     void *user_data) GRPC_MUST_USE_RESULT;
 
+/// Remove any filter whose name is \a filter_name from \a builder. Returns true
+/// upon success.
```
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.h]: s/upon success/if the specified filter name was found/

File: test/core/channel/channel_stack_builder_test.c [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+      grpc_insecure_channel_create("target name isn't used", NULL, NULL);
+  GPR_ASSERT(channel != NULL);
+  // Make sure the high priority filter has been created.
+  GPR_ASSERT(g_high_prio_arg == 0xdeadbeef);
+  // ... and that the low priority one hasn't.
+  GPR_ASSERT(!g_low_prio_fn_called);
+  grpc_channel_destroy(channel);
+}
+
+const grpc_channel_filter filter_high_prio = {
```
Comment (markdroth) [inline test/core/channel/channel_stack_builder_test.c]: Suggest calling this replacement_filter.

File: test/core/channel/channel_stack_builder_test.c [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    grpc_call_stack_ignore_set_pollset_or_pollset_set,
+    call_destroy_func,
+    0,
+    channel_init_func,
+    channel_destroy_func,
+    get_peer,
+    grpc_channel_next_get_info,
+    "filter_name"};
+
+const grpc_channel_filter filter_low_prio = {
```
Comment (markdroth) [inline test/core/channel/channel_stack_builder_test.c]: Suggest calling this original_filter.

File: test/core/channel/channel_stack_builder_test.c [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    grpc_call_stack_ignore_set_pollset_or_pollset_set,
+    call_destroy_func,
+    0,
+    channel_init_func,
+    channel_destroy_func,
+    get_peer,
+    grpc_channel_next_get_info,
+    "filter_name"};
+
+static bool add_high_prio_filter(grpc_exec_ctx *exec_ctx,
```
Comment (markdroth) [inline test/core/channel/channel_stack_builder_test.c]: Suggest calling this replace_filter().

File: test/core/channel/channel_stack_builder_test.c [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+                                 void *arg) {
+  const grpc_channel_filter *filter = arg;
+  // Get rid of any other version of the filter, as determined by having the
+  // same name.
+  GPR_ASSERT(grpc_channel_stack_builder_remove_filter(builder, filter->name));
+  return grpc_channel_stack_builder_prepend_filter(
+      builder, filter, high_prio_fn, &g_high_prio_arg);
+}
+
+static bool add_low_prio_filter(grpc_exec_ctx *exec_ctx,
```
Comment (markdroth) [inline test/core/channel/channel_stack_builder_test.c]: Suggest calling this add_original_filter().

File: test/core/channel/channel_stack_builder_test.c [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+    GRPC_ERROR_UNREF(op->disconnect_with_error);
+  }
+  GRPC_CLOSURE_SCHED(exec_ctx, op->on_consumed, GRPC_ERROR_NONE);
+}
+
+static char *get_peer(grpc_exec_ctx *exec_ctx, grpc_call_element *elem) {
+  return gpr_strdup("peer");
+}
+
+unsigned long g_high_prio_arg = 0;
```
Comment (markdroth) [inline test/core/channel/channel_stack_builder_test.c]: Couldn't this just be a bool, similar to what you're doing for the original filter?

File: src/core/lib/channel/channel_stack_builder.h (hunk 7 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack_builder.h]: Done

File: src/core/lib/channel/channel_stack_builder.h (hunk 10 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack_builder.h]: Done

File: src/core/lib/channel/channel_stack_builder.c (hunk 9 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack_builder.c]: Good suggestion. Done.

File: src/core/lib/channel/channel_stack_builder.c (hunk 8 above)
Comment (dgquintas) [inline src/core/lib/channel/channel_stack_builder.c]: In fact it'll always be NULL the first iteration. The first element of the list `it` iterates over (ie, `builder->begin`) is a market node with its filter field always set to NULL. We could in fact do a `next(it)` after calling `grpc_channel_stack_builder_create_iterator_at_first(builder)` but it looks ugly.

File: test/core/channel/channel_stack_builder_test.c (hunk 15 above)
Comment (dgquintas) [inline test/core/channel/channel_stack_builder_test.c]: Yeah. But where would I display my l33tspeak!

d0n3.

File: test/core/channel/channel_stack_builder_test.c (hunk 11 above)
Comment (dgquintas) [inline test/core/channel/channel_stack_builder_test.c]: Done

File: test/core/channel/channel_stack_builder_test.c (hunk 12 above)
Comment (dgquintas) [inline test/core/channel/channel_stack_builder_test.c]: Done

File: test/core/channel/channel_stack_builder_test.c (hunk 13 above)
Comment (dgquintas) [inline test/core/channel/channel_stack_builder_test.c]: Done

File: test/core/channel/channel_stack_builder_test.c (hunk 14 above)
Comment (dgquintas) [inline test/core/channel/channel_stack_builder_test.c]: Done

File: src/core/lib/channel/channel_stack_builder.c (hunk 8 above)
Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.c]: Oh!  In that case, I suggest replacing the for loop with a while loop:

```
while (grpc_channel_stack_builder_move_next(it)) {
```

Then there's no need to check for the filter name being NULL.  And it's simpler, to boot.

File: src/core/ext/filters/load_reporting/load_reporting.c [hunk 16]
```diff
@@ -42,9 +42,13 @@ static bool maybe_add_load_reporting_filter(grpc_exec_ctx *exec_ctx,
...
   const grpc_channel_args *args =
       grpc_channel_stack_builder_get_channel_arguments(builder);
-  if (is_load_reporting_enabled(args)) {
-    return grpc_channel_stack_builder_prepend_filter(
-        builder, (const grpc_channel_filter *)arg, NULL, NULL);
+  const grpc_channel_filter *filter = arg;
+  grpc_channel_stack_builder_iterator *it = grpc_channel_stack_builder_iterator_find(builder, filter->name);
+  const bool already_has_load_reporting_filter = !grpc_channel_stack_builder_iterator_is_end(it);
+  grpc_channel_stack_builder_iterator_destroy(it);
+  if (is_load_reporting_enabled(args) && !already_has_load_reporting_filter) {
```
Comment (dgquintas) [inline src/core/ext/filters/load_reporting/load_reporting.c]: The test's initial call to `::testing::InitGoogleTest()` will trigger all (internal) module initializers. One of them is `InitGRPC()`, which will register the internal version of the load reporting plugin.

The OSS version, from `grpc_init()` is, in this case, initialized later as part of instantiating `ServerCompletionQueue` because [`CompletionQueue` inherits from `GrpcLibraryCodegen`](https://github.com/grpc/grpc/blob/f4af6eb2f44e7a9fa08e5e427a5e5f9d1ba2acd2/include/grpc%2B%2B/impl/codegen/completion_queue.h#L86), whose job is precisely to [call `grpc_init()`](https://github.com/grpc/grpc/blob/5253c8f9a899450397a5e46e4923d01ac9a66a27/include/grpc%2B%2B/impl/codegen/grpc_library.h#L44). `ServerCompletionQueue` is not instantiated until the test needs to create a server instance.

Basically, internal happens first because the (internal) testing framework initializes internal modules first.

**Comment by dgquintas** [comment]
jenkins test this please

**Comment by dgquintas** [comment]
Issues: #12261 #12435
